### PR TITLE
add TypeDoc to project

### PR DESCRIPTION
**Pull request by akash1810**
## What does this change?

[TypeDoc](https://typedoc.org/) will generate a documentation site which we can publish to GitHub Pages.

A new `script/docs` script is added to view docs locally and `script/release` is updated to run [`gh-pages`](https://www.npmjs.com/package/gh-pages) to publish the generated site to GitHub Pages.

## Does this change require changes to existing projects or CDK CLI?

No.

## How to test

- Checkout branch
- Run `script/docs`

## How can we measure success?

Published documentation 🎉 .

This will also encourage us to annotate our code properly, as noted [here](https://github.com/guardian/cdk/pull/137#discussion_r556698030).

## Have we considered potential risks?

n/a

### Index Page
image

### Example page for `GuStack`
image

### Example page for `GuStack` w/out externals (AWS CDK properties) shown
image

### CODE REVIEW

**Review by jamie-lynch — COMMENTED** [review]
This is great! 📚

File: script/release [hunk 1]
```diff
@@ -9,6 +9,7 @@ DIR=$( cd "$( dirname "${BASH_SOURCE[0]}" )" && pwd )
 if npm whoami &>/dev/null; then
   rm -rf ./lib
   npm run release
+  npm run release:docs
```
[the remark this answers is not in the document]
Comment (akash1810) [reply script/release:12]: Oh great point! Thoughts on possible solutions?

- Have a separate script and remember to run it
- Publish to GH Pages automatically via GH Actions, triggered by the creation of a GH tag/release
- `npm run release || true` to always have that statement exit cleanly
- Something else

[the remark this answers is not in the document]
Comment (jamie-lynch) [reply script/release:12]: I'd vote for the second one as:
* I will forget to run a separate script
* If it fails for another reason (like yesterday when the tests failed) we probably don't want to publish to docs

[the remark this answers is not in the document]
Comment (akash1810) [reply script/release]: Cool. Added a GH Actions workflow file now, which _should_ deploy documentation to GH Pages whenever a new release is created.

File: typedoc.js [hunk 2]
```diff
@@ -0,0 +1,19 @@
+const { readdirSync } = require("fs");
+
+const constructsDir = "src/constructs";
+
+module.exports = {
+  entryPoints: [
+    "src/index.ts",
```
[the remark this answers is not in the document]
Comment (akash1810) [reply typedoc.js:7]: Done.

**Review by jamie-lynch — APPROVED** [review]

**Pull request merged**